Meetboutengis

Vegter opened the pull request:
Solves issue #6 

Importclient for meetboutengis - import meetbouten

Vegter commented:
This branch is derived from client-basic-structure

spreeker commented:
Meetbouten? Da's wel een heel saaie test Dataset. Die verandert niet meer afaik :)

dolhuis commented:
> What was wrong with using abstract base class? I do not see any comment in the commit message. The generic apporach was to import each set without knowledge about it. Every set that implements the generic interface can be imported. Please note that the other meetbouten datasets also need to be imported, that's why an abstract base class was used

I don't think there is something wrong with abstract base classes as such. I don't see the fit here, and when I separated concerns the need to use it disappeared. The code is now complete generic, so you can import each set without knowledge about it. And I guess that the other meetbouten datasets can now be imported without any additional code, let alone an expanded class hierarchy.

I hope you can see merit in this way of organising the code (generic code and configuration per dataset) over the class-hierarchy (in which specific code was mixed with knowledge about the dataset).

The more generic the code is upfront, the more generic it can be downstream.
The more the code is dataset-agnostic the higher the possibility of reuse.

> I suggest adding comments why using schema's. Now it is a little bit magic...

That was also why I added the schema, but that is a very immature in this version. So I guess you're right, there.

I imagine the possibility to tie implementations (for instance the - currently - ugly `if type == 'string' ... elif type == 'decimal' ` etc. ) with the configuration. That would allow for an almost automatic mapping of generic bits of code to the appropriate datasets, entities or fields. And that, in turn, could be used in other components of the system as well.

Vegter commented:
Ok, let’s see when the other meetbouten datasets gets imported. My goal was
to be independent of the dataset and to run an import without knowing what
is being imported. That was the reason why I introduced an array of
imports. Each import is processed identically because it implements the
genieric interface. But I am positive to any better solution. Top.

On Tue, 31 Jul 2018 at 19:11, dolhuis <notifications@github.com> wrote:

> What was wrong with using abstract base class? I do not see any comment in
> the commit message. The generic apporach was to import each set without
> knowledge about it. Every set that implements the generic interface can be
> imported. Please note that the other meetbouten datasets also need to be
> imported, that's why an abstract base class was used
>
> I don't think there is something wrong with abstract base classes as such.
> I don't see the fit here, and when I separated concerns the need to use it
> disappeared. The code is now complete generic, so you can import each set
> without knowledge about it. And I guess that the other meetbouten datasets
> can now be imported without any additional code, let alone an expanded
> class hierarchy.
>
> I hope you can see merit in this way of organising the code (generic code
> and configuration per dataset) over the class-hierarchy (in which specific
> code was mixed with knowledge about the dataset).
>
> The more generic the code is upfront, the more generic it can be
> downstream.
> The more the code is dataset-agnostic the higher the possibility of reuse.
>
> I suggest adding comments why using schema's. Now it is a little bit
> magic...
>
> That was also why I added the schema, but that is a very immature in this
> version. So I guess you're right, there.
>
> I imagine the possibility to tie implementations (for instance the -
> currently - ugly if type == 'string' ... elif type == 'decimal' etc. )
> with the configuration. That would allow for an almost automatic mapping of
> generic bits of code to the appropriate datasets, entities or fields. And
> that, in turn, could be used in other components of the system as well.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Amsterdam/GOB-Import-Client-Template/pull/7#issuecomment-409297287>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGXtwsE_KxXaWyPVJgj4F4qkS8AIaWhIks5uMI-2gaJpZM4VQ37A>
> .
>
-- 

Met vriendelijke groet,

Jurien Vegter

dolhuis commented:
It is good to ask for my rationale, it forces me to express why I think this is the way to go. And also ideas should be tested, not only code :-)

Some more thoughts:
- if we have an implementation-class per dataset, one could be more easily tempted to add one-off code that will only work in that particular situation.
- conversely, if we have a generic set of code, and you have to configure the use of it, we could be stimulated to reason on a more abstract level on the process of reading and converting data, preparing it for import

dolhuis commented:
@Vegter If you have the time, please review the current state of this branch and if okay, let me know, I will merge it.

dolhuis commented:
Just a last refinement, shall we merge it then?

Vegter commented:
Agreed

dolhuis commented:
> Agreed

OK. Will do so locally and push it to `origin`, due to changes previously merged/rebased to `develop` a rebase and merge of this pull-request is no longer possible.
